micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmGcMetrics.java

```diff
@@ -133,64 +136,74 @@ public void bindTo(MeterRegistry registry) {
+                                    if (delta > 0L) {
+                                        promotedBytes.increment(delta);
+                                    }
+
+                                    // Some GC implementations such as G1 can reduce the old gen size as part of a minor GC. To track the
+                                    // live data size we record the value if we see a reduction in the old gen heap size or
+                                    // after a major GC.
+                                    if (oldAfter < oldBefore || GcGenerationAge.fromName(notificationInfo.getGcName()) == GcGenerationAge.OLD) {
+                                        liveDataSize.set(oldAfter);
+                                        final long oldMaxAfter = after.get(oldGenPoolName).getMax();
+                                        maxDataSize.set(oldMaxAfter);
+                                    }
                                 }
 
-                                // Some GC implementations such as G1 can reduce the old gen size as part of a minor GC. To track the
-                                // live data size we record the value if we see a reduction in the old gen heap size or
-                                // after a major GC.
-                                if (oldAfter < oldBefore || GcGenerationAge.fromName(notificationInfo.getGcName()) == GcGenerationAge.OLD) {
-                                    liveDataSize.set(oldAfter);
-                                    final long oldMaxAfter = after.get(oldGenPoolName).getMax();
-                                    maxDataSize.set(oldMaxAfter);
+                                if (youngGenPoolName != null) {
+                                    final long youngBefore = before.get(youngGenPoolName).getUsed();
+                                    final long youngAfter = after.get(youngGenPoolName).getUsed();
+                                    final long delta = youngBefore - youngGenSizeAfter.get();
+                                    youngGenSizeAfter.set(youngAfter);
+                                    if (delta > 0L) {
+                                        allocatedBytes.increment(delta);
+                                    }
                                 }
-                            }
 
-                            if (youngGenPoolName != null) {
-                                final long youngBefore = before.get(youngGenPoolName).getUsed();
-                                final long youngAfter = after.get(youngGenPoolName).getUsed();
-                                final long delta = youngBefore - youngGenSizeAfter.get();
-                                youngGenSizeAfter.set(youngAfter);
-                                if (delta > 0L) {
-                                    allocatedBytes.increment(delta);
+                                try {
+                                    ((NotificationEmitter) mbean).removeNotificationListener(this);
```

Comment:
Maybe my diff viewing skills are fooling me, but isn't this removing the listener exactly after the first notification?

Reply:
@mweirauch I thought the first notification is enough for binding meters similar to Tomcat one but didn't check. Are the subsequent events considered to get meters bound fully?

Reply:
@mweirauch Oops. Looking into the code closely, this implementation seems completely wrong. Thanks for pointing this out.

Reply:
Unless there's a proper shutdown hook, I'm not sure how to remove the listener at the moment.

Reply:
One way might be to make `JvmGcMetrics` `AutoCloseable`.

Reply:
@mweirauch Thanks for the feedback! I changed `JvmGcMetrics` to implement `AutoCloseable` although I'm not sure there's a better fix for this.